src/test/java/org/kohsuke/github/extras/okhttp3/GitHubCachingTest.java

```diff
@@ -0,0 +1,334 @@
+    @Test
+    public void OkHttpConnector_Cache_MaxAgeDefault_Zero_GitHubRef_Error() throws Exception {
+
+        // requireProxy("For clarity. Will switch to snapshot shortly.");
+        // snapshotNotAllowed();
+
+        OkHttpClient client = createClient(true);
+        OkHttpConnector connector = new OkHttpConnector(client);
+
+        this.gitHub = getGitHubBuilder().withEndpoint(mockGitHub.apiServer().baseUrl())
+                .withConnector(connector)
+                .build();
+
+        // Alternate client also doing caching but staying in a good state
+        // We use this to do sanity checks and other information gathering
+        GitHub gitHub2 = getGitHubBuilder().withEndpoint(mockGitHub.apiServer().baseUrl())
+                .withConnector(new OkHttpConnector(createClient(true)))
+                .build();
+
+        // Create a branch from a known conflicting branch
+        GHRepository repo = getRepository(gitHub);
+
+        String baseSha = repo.getRef("heads/test/unmergeable").getObject().getSha();
+
+        GHRef ref;
+        ref = repo.createRef("refs/" + testRefName, baseSha);
+
+        // Verify we can query the created ref
+        ref = repo.getRef(testRefName);
+
+        // Verify we can query the created ref from cache
+        ref = repo.getRef(testRefName);
+
+        // Delete the ref
+        ref.delete();
+
+        // This is just to show this isn't a race condition
+        Thread.sleep(2000);
+
+        // Try to get the non-existant ref (GHFileNotFound)
+        try {
+            repo.getRef(testRefName);
+            fail();
+        } catch (GHFileNotFoundException e) {
+            // expected
+
+            // FYI: Querying again when the item is actually not present does not produce a 304
+            // It produces another 404,
+            // Try to get the non-existant ref (GHFileNotFound)
+            try {
+                repo.getRef(testRefName);
+                fail();
+            } catch (GHFileNotFoundException ex) {
+                // expected
+            }
+
+        }
+
+        // This is just to show this isn't a race condition
+        Thread.sleep(2000);
+
+        ref = repo.createRef("refs/" + testRefName, baseSha);
+
+        // Verify ref exists and can be queried from uncached connection
+        // Expected: success
+        // Actual: still GHFileNotFound due to caching: GitHub incorrectly returns 304
+        // even though contents of the ref have changed.
+        //
+        // There source of this issue seems to be that 404's do not return an ETAG,
+        // so the cache falls back to using "If-Modified-Since" which is erroneously returns a 304.
+        //
+        // NOTE: This is even worse than you might think: 404 responses don't return an ETAG, but 304 responses do.
+        //
+        // Due erroneous 304 returned from "If-Modified-Since", the ETAG returned by the first 304
```

Comment:
This refers to https://github.com/github-api/github-api/pull/665/files#diff-02b86edbc1ffd10d08e7049e7a842fc8R9.